sql/wsrep_mysqld.cc

```diff
@@ -1502,6 +1502,44 @@ static bool wsrep_can_run_in_toi(THD *thd, const char *db, const char *table,
     {
       return false;
     }
+    /*
+      If mariadb master has replicated a CTAS, we should not replicate the create table
+      part separately as TOI, but to replicate both create table and following inserts
+      as one write set.
+      Howver, if CTAS creates empty table, we should replicate the create table alone
+      as TOI. We have to do relay log event lookup to see if row events follow the
+      create table event.
+    */
+    if (thd->slave_thread && !(thd->rgi_slave->gtid_ev_flags2 & Gtid_log_event::FL_STANDALONE))
+    {
+      /* this is CTAS, either empty or populated table */
+      ulonglong event_size = 0;
+      Log_event* ev = wsrep_peak_event(thd->rgi_slave, &event_size);
+      if (ev)
+      {
+	switch (ev->get_type_code())
+	{
+	case QUERY_EVENT:
+	  /* CTAS with empty table, we replicate create table as TOI */
+	  break;
+
+        case TABLE_MAP_EVENT:
+          WSREP_DEBUG("replicating CTAS of empty table as TOI");
+          // fall through
```

Comment:
Is this correct, in case of TABLE_MAP_EVENT we are executing also code for WRITE_ROWS_EVENTS?